Closes #764: Add 'defaultSession' to SessionManager

grigoryk opened the pull request:
Fixed #764. Split from #826.

pocmo commented on components/browser/session/src/main/java/mozilla/components/browser/session/SessionManager.kt:
```diff
@@ -173,18 +181,31 @@ class SessionManager(
+            add(defaultSession.invoke())
+        } else {
+            notifyObservers { onAllSessionsRemoved() }
+        }
```
That's the part I'm not sure about: An observer would never see the removal, just an add. Why wouldn't we first notify about removal and then add the new session?

grigoryk replied:
We could! With a default session present, we're never really in an "empty" state that's observable to the outside world - that's why I didn't notify observers. Reflecting on the docstring of "all sessions removed" callback, it seems that the intent there isn't to indicate an empty manager, but to indicate that one of the blanket remove* functions were invoked. That interpretation would dictate that yeah, they should be notified here.

grigoryk replied:
ISTM that the observer "onAllSessionsRemoved" method should be named differently, since it doesn't actually fire when "all sessions are removed". Filed a follow up #948.

csadilek commented:
@grigoryk the test `default session not used when all sessions were removed and they were all CustomTab` seems to have failed on Travis.

grigoryk commented:
@csadilek not sure why, I think I forgot to push something since it is passing locally. I've addressed @pocmo question and changed the behaviour to fire the "onAllSessionsRemoved" callbacks, and filed a follow-up to clarify this (I think it's a confusing API). See #948.

... and now TC is failing for entirely unrelated reasons :-)
